geopandas/tools/overlay.py

```diff
@@ -176,3 +176,143 @@ def overlay(df1, df2, how, use_sindex=True):
+    -------
+    df : GeoDataFrame
+        GeoDataFrame with new set of polygons and attributes
+        resulting from the overlay
+
+    """
+    if use_sindex is not None:
+        print('use_sindex is deprecated. If you are trying to use the old "overlay" function use "overlay_slow".')
+
+    # Allowed operations
+    allowed_hows = [
+        'intersection',
+        'union',
+        'identity',
+        'symmetric_difference',
+        'difference',  # aka erase
+    ]
+    # Error Messages
+    if how not in allowed_hows:
+        raise ValueError("`how` was \"%s\" but is expected to be in %s" % \
+            (how, allowed_hows))
+
+    if isinstance(df1, GeoSeries) or isinstance(df2, GeoSeries):
+        raise NotImplementedError("overlay currently only implemented for GeoDataFrames")
+
+    if (df1.geom_type.apply(lambda x: x in ['Polygon', 'MultiPolygon']).sum()!=len(df1.index) or 
+        df2.geom_type.apply(lambda x: x in ['Polygon', 'MultiPolygon']).sum()!=len(df2.index)):
+        raise TypeError("overlay only takes GeoDataFrames with (multi)polygon geometries") 
+
+    # Computations
+    df1 = df1.copy()
+    df2 = df2.copy()
+    df1['geometry'] = df1.geometry.buffer(0)
+    df2['geometry'] = df2.geometry.buffer(0)
+    if df1.crs!=df2.crs and reproject:
+        print('Data has different projections.')
+        print('Converted data to projection of first GeoPandas DatFrame')
+        df2.to_crs(crs=df1.crs, inplace=True)
+    if how=='intersection':
+        # Spatial Index to create intersections
+        spatial_index = df2.sindex
+        df1['bbox'] = df1.geometry.apply(lambda x: x.bounds)
+        df1['sidx']=df1.bbox.apply(lambda x:list(spatial_index.intersection(x)))
+        pairs = df1['sidx'].to_dict()
+        nei = []
+        for i,j in pairs.items():
+            for k in j:
+                nei.append([i,k])
+        if nei!=[]:
+            pairs = GeoDataFrame(nei, columns=['idx1','idx2'], crs=df1.crs)
+            pairs = pairs.merge(df1, left_on='idx1', right_index=True)
+            pairs = pairs.merge(df2, left_on='idx2', right_index=True, suffixes=['_1','_2'])
+            pairs['Intersection'] = pairs.apply(lambda x: (x['geometry_1'].intersection(x['geometry_2'])).buffer(0), axis=1)
+            pairs = GeoDataFrame(pairs, columns=pairs.columns, crs=df1.crs)
+            cols = pairs.columns.tolist()
+            cols.remove('geometry_1')
+            cols.remove('geometry_2')
+            cols.remove('sidx')
+            cols.remove('bbox')
+            cols.remove('Intersection')
+            dfinter = pairs[cols+['Intersection']].copy()
+            dfinter.rename(columns={'Intersection':'geometry'}, inplace=True)
+            dfinter = GeoDataFrame(dfinter, columns=dfinter.columns, crs=pairs.crs)
+            dfinter = dfinter.loc[dfinter.geometry.is_empty==False]
+            #dfinter.drop(['idx1','idx2'], inplace=True, axis=1)
+            dfinter.reset_index(inplace=True, drop=True)
```

Comment:
In general, we don't use the inplace style within geopandas, so I would prefer to write `dfinter = dfinter.reset_index(drop=True)`

Reply:
OK. Does this apply to any change that uses ``inplace=True``? E.g. ``to_crs``?